lsp/server/src/server.ts

```diff
@@ -34,92 +34,85 @@ const documents: TextDocuments<TextDocument> = new TextDocuments(TextDocument);
-		// Register for all configuration changes.
-		connection.client.register(DidChangeConfigurationNotification.type, undefined);
-		connection.workspace.getConfiguration('SA4U').then((value) => {
-			if (value.messageDefinition) {
-				messDef = value.messageDefinition;
-			}
-		});
-	}
-	if (hasWorkspaceFolderCapability) {
-		connection.workspace.onDidChangeWorkspaceFolders(_event => {
-			connection.console.log('Workspace folder change event received.');
-		});
-	}
-	connection.workspace.getConfiguration();
+async function checkForDockerRunningAndInstalled() {
+	return await execAsync('docker info')
+		.then(() => {return Promise.resolve;}, () => {return Promise.reject('Please verify Docker is installed and Docker daemon is running.');});
+}
+
+async function startDockerContainer() {
 	connection.workspace.getWorkspaceFolders().then((folders:null|WorkspaceFolder[]|void) => {
 		if (folders) {
 			folders.forEach((folder) => {
 				dockerContainer(folder);
 			});
 		}
 	});
-	documents.all().forEach((doc)=>validateTextDocument(doc));
-});
-
-class SA4UConfig {
-	compilationDir: string
-	constructor(compilationDir: string) {
-		this.compilationDir = '/src/' + compilationDir;
-	}
 }
 
-// TODO: it would be nice to automatically discover the compile_commands.json file.
-const readFileAsync = promisify(readFile);
+async function stopDockerContainer() {
+	const output = await execAsync(`docker container ls -q -f name=sa4u_z3_server`)
+		.then((value) => {return value.stdout.split('\n');})
+		.catch((err) => {console.log(err); return [];});
+	await execAsync(`docker container stop ${output.join(' ')}`);
+}
 
-async function getSA4UConfig(projectRootDir: string) {
-	try {
-		const configJSON = JSON.parse((await readFileAsync(projectRootDir + "/" + ".sa4u.json")).toString());
-		return new SA4UConfig(configJSON['CompilationDir']);
-	} catch (err) {
-		console.warn('unable to read SA4U config: ' + err);
+async function getSA4UConfig() {
+	const readConfig = await promises.readFile(".sa4u.json")
+		.then((value)=>{return value.toString();})
+		.catch((err)=>{console.warn('unable to read SA4U config: ' + err); return new SA4UConfig('');});
+	if (readConfig instanceof SA4UConfig) {
+		return readConfig;
+	}
+	const configJSON = JSON.parse(readConfig);
+	if (typeof(configJSON['CompilationDir'])=='undefined') {
 		return new SA4UConfig('');
 	}
+	if(!existsSync('.sa4u')){
+		await promises.mkdir('.sa4u')
+			.catch((err) => {console.warn(`Failed to create .sa4u directory: ${err}`); return new SA4UConfig(configJSON['CompilationDir']);});
+	}
+	const directory = resolve("./");
+	const readCompileCommands = await promises.readFile(configJSON['CompilationDir']+"/compile_commands.json")
+		.then((value) => {return value.toString().replace(directory, '/src/');})
+		.catch((err) => {console.warn(`Failed to read ${configJSON['CompilationDir']}/compile_commands.json: ${err}`); return new SA4UConfig('');});
+	if(readCompileCommands instanceof SA4UConfig)
```

Comment:
Add a space after the `if` and the condition

Reply:
Fixed